finat/mappers.py

```diff
@@ -188,3 +191,40 @@ def map_index_sum(self, expr, *args, **kwargs):
 
 class GraphvizMapper(WalkMapper, GVM):
     pass
+
+
+class BindingMapper(IdentityMapper):
+    """A mapper that binds free indices in recipes using ForAlls."""
+
+    def __init__(self, kernel_data):
+        """
+        :arg context: a mapping from variable names to values
+        """
+        super(BindingMapper, self).__init__()
+        self.bound_above = {}
+        self.bound_below = {}
```

Comment:
Is there a reason to use dictionaries here rather than sets? As far as I can see you always set the values in the dictionary to None so you are effectively using the dictionary as a set of keys.